hathor/dag_builder/builder.py

```diff
@@ -0,0 +1,180 @@
+                    self.add_spending_edge(_from, _to, _txout_index)
+
+                case (TokenType.ATTRIBUTE, (name, key, value)):
+                    self.add_attribute(name, key, value)
+
+                case (TokenType.ORDER_BEFORE, (_from, _to)):
+                    self.add_deps(_from, _to)
+
+                case (TokenType.OUTPUT, (name, index, amount, token, attrs)):
+                    self.set_output(name, index, amount, token, attrs)
+
+                case (TokenType.BLOCKCHAIN, (name, first_parent, begin_index, end_index)):
+                    self.add_blockchain(name, first_parent, begin_index, end_index)
+
+                case _:
+                    raise NotImplementedError(parts)
+
+    def _get_node(self, name: str) -> DAGNode:
+        return self._nodes[name]
+
+    def _get_or_create_node(self, name: str, *, default_type: str = 'unknown') -> DAGNode:
+        if name not in self._nodes:
+            self._nodes[name] = DAGNode(name=name, type=default_type)
+        return self._nodes[name]
+
+    def add_deps(self, _from: str, _to: str) -> Self:
+        from_node = self._get_or_create_node(_from)
+        self._get_or_create_node(_to)
+        from_node.deps.add(_to)
+        return self
+
+    def add_blockchain(self, prefix: str, first_parent: str | None, first_index: int, last_index: int) -> Self:
+        prev = first_parent
+        for i in range(first_index, last_index + 1):
+            name = f'{prefix}{i}'
+            self._get_or_create_node(name, default_type='block')
+            if prev is not None:
+                self.add_parent_edge(name, prev)
+            prev = name
+        return self
+
+    def add_parent_edge(self, _from: str, _to: str) -> Self:
+        self._get_or_create_node(_to, default_type='transaction')
+        from_node = self._get_or_create_node(_from, default_type='transaction')
+        from_node.parents.add(_to)
+        return self
+
+    def add_spending_edge(self, _from: str, _to: str, _txout_index: int) -> Self:
+        self._get_or_create_node(_to, default_type='transaction')
+        from_node = self._get_or_create_node(_from, default_type='transaction')
+        from_node.inputs.add(DAGInput(_to, _txout_index))
+        return self
+
+    def set_output(self, name: str, index: int, amount: int, token: str, attrs: AttributeType) -> Self:
+        node = self._get_or_create_node(name)
+        if len(node.outputs) <= index:
+            node.outputs.extend([None] * (index - len(node.outputs) + 1))
+        node.outputs[index] = DAGOutput(amount, token, attrs)
+        if token != 'HTR':
+            self._get_or_create_node(token, default_type='token')
+            node.deps.add(token)
+        return self
+
+    def add_attribute(self, name: str, key: str, value: str) -> Self:
+        node = self._get_or_create_node(name)
+        node.attrs[key] = value
+        return self
+
+    def topological_sorting(self) -> Iterator[DAGNode]:
```

Comment:
Can't you reuse the `traversal` module? I mean, by making necessary adaptations.

Reply:
No, because the traversal module implements different types of BFS and DFS. Here I'm using Kahn's algorithm.